src/librbd/mirror/snapshot/UnlinkPeerRequest.cc

```diff
@@ -91,15 +91,20 @@ void UnlinkPeerRequest<I>::unlink_peer() {
     finish(-EINVAL);
     return;
   }
-
-  // if there is or will be no more peers in the mirror snapshot and we have
-  // a more recent mirror snapshot, remove the older one
   if ((mirror_ns->mirror_peer_uuids.count(m_mirror_peer_uuid) == 0) ||
       (mirror_ns->mirror_peer_uuids.size() <= 1U && m_newer_mirror_snapshots)) {
```

Comment:
I can't see a scenario where it can be less than 1.

Reply:
Actually I am not sure I understand that condition. According to the comment:
```
  // if there is or will be no more peers in the mirror snapshot and we have
  // a more recent mirror snapshot, remove the older one
```
The condition should be:
```
if ((mirror_ns->mirror_peer_uuids.size() == 0 ||
     mirror_ns->mirror_peer_uuids.size() == 1 && mirror_ns->mirror_peer_uuids.count(m_mirror_peer_uuid) != 0) && 
     m_newer_mirror_snapshots) {
```
And actually it looks like it is always safe to remove the snapshot if there are no peers already (even if there are no newer snapshots), so it could probably be simplified to:
```
if (mirror_ns->mirror_peer_uuids.size() == 0 ||
    mirror_ns->mirror_peer_uuids.size() == 1 && mirror_ns->mirror_peer_uuids.count(m_mirror_peer_uuid) != 0 && m_newer_mirror_snapshots) {
```
But I don't see why we would want to always remove a snapshot that has no our peer (as it is in the current code). It will not work for multipeer. Probably it is because we do not support multipeer currently and consider all other peers as invalid. There may be a bug here if the multipeer support is added though.

Reply:
> And actually it looks like it is always safe to remove the snapshot if there are no peers already (even if there are no newer snapshots)

I don't think that is the case -- we would lose all snapshot-based mirroring information for the image (whether the image is primary, etc).

> But I don't see why we would want to always remove a snapshot that has no our peer (as it is in the current code).

I think you fell into the confusion trap I highlighted in another comment:

> * Why are the checks for whether the snapshot should _actually_ be removed split between `unlink_peer()` and `remove_snapshot()` in the first place?  The fact that `remove_snapshot()` can skip removal of snapshot causes nothing but confusion.

`remove_snapshot()` wouldn't actually remove a snapshot that has no our peer ;)  This code clearly needs to be refactored.